contrib/dhcp_agent/dhcp_agent.py

```diff
@@ -0,0 +1,438 @@
+        should_restart_dhcp_server = False
+        successful_list = []
+        for section in self.sections:
+            is_saved = False
+            dhcp_config = self._get_configuration(section, request_params)
+            if dhcp_config:
+                is_saved = self._set_new_configuration(
+                    dhcp_config, self.section_config_path_mapper[section]
+                )
+                successful_list.append(is_saved)
+        should_restart_dhcp_server = any(successful_list)
+
+        if self.can_restart_dhcp_server and should_restart_dhcp_server:
+            self._restart_dhcp_server()
+        self._send_synch_confirmation()
+
+    def make_authorized_request(self, url, params=None):
+        """Make request with extra headers like Authorization
+        and If-Modified-Since.
+
+        Headers sending with request:
+            - Authorization: contains token to autorization on the server,
+            - If-Modified-Since: information for server about configuration's
+                last date.
+
+        Returns:
+            object: standard response object (file-like object)
+        """
+        headers = {}
+        if params:
+            params = urlencode(params)
+            url += '?' + params
+        last = self.cache.get(prefix=CACHE_LAST_MODIFIED_PREFIX, key=url)
+        if last:
+            headers['If-Modified-Since'] = last
+        headers.update({'Authorization': 'Token {}'.format(self.key)})
+        return urlopen(Request(url, headers=headers))
+
+    def _get_configuration(self, mode, params):
+        """Fetching configuration from Ralph server."""
+        url = '{}://{}/dhcp/{}/'.format(
+            self.proto,
+            self.host,
+            mode,
+        )
+        configuration = None
+
+        self.logger.info('Sending request to {}'.format(url))
+        try:
+            response = self.make_authorized_request(url, params)
+        except HTTPError as e:
+            if e.code != 304:
+                self.logger.error(
+                    'Server returned {} status code with message "{}"'.format(
+                        e.code, e.fp.read().decode()
+                    )
+                )
+            else:
+                self.logger.info(
+                    'Server return status 304 NOT MODIFIED. Nothing to do.'
+                )
+            return False
+        else:
+            configuration = response.read()
+            self.cache.set(
+                prefix=CACHE_LAST_MODIFIED_PREFIX,
+                key=url,
+                value=response.headers.get('Last-Modified')
+            )
+        return configuration
+
+    def _send_synch_confirmation(self):
+        """This method notify Ralph about restarting DHCP server with
+        new configuration.
+
+        Returns:
+            bool: True if server returns 200 status code, otherwise False
+        """
+        url = '{}://{}/dhcp/synch/'.format(
```

Comment:
Maybe it's just me, but `synch` looks kind of strange in this context - what about `sync`..?